server/controllers/activitypub/lib/shared_key_pair.js

```diff
@@ -7,12 +7,14 @@
 // cached keys will get refreshed at their next attempt
 // See https://socialhub.activitypub.rocks/t/caching-public-keys/688
 
-const { generateKeyPair } = require('lib/crypto').keyPair
+const { generateRsaKeyPair, sha1 } = require('lib/crypto')
 
 let sharedKeyPair
 
 const getSharedKeyPair = async () => {
-  sharedKeyPair = sharedKeyPair || await generateKeyPair()
+  if (sharedKeyPair) return sharedKeyPair
+  sharedKeyPair = await generateRsaKeyPair()
+  sharedKeyPair.publicKeyHash = sha1(sharedKeyPair.publicKey).slice(0, 10)
```

Comment:
why sha1 and not just a slice, or even the `publicKey` itself ?

Reply:
`sha1` is fast and this hashing is done only once per server process, while `publicKey` looks like `"-----BEGIN PUBLIC KEY-----\nMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCg [...]` , and would need some transformation to be nice in a URL anyway